`python3Packages.jaxlibWithCuda` build failure on x86_64-linux as of `891d0226`

## Issue description
Build of `python3Packages.jaxlib` failed on x86_64-linux as of 891d0226de6e281422a8d0d99814925fd7aed8d0. This is currently breaking python3Packages.jaxlibWithCuda.

```
error: builder for '/nix/store/3dx5b3blc4girz6zf772vqrsgw8457z3-bazel-build-jaxlib-0.3.0.drv' failed with exit code 1;
       last 10 log lines:
       > /nix/store/m2vh2ny7bqpwij1gpmvl5gxj7y4dgr4f-binutils-2.38/bin/ld: /nix/store/x0gmnlsh4mzfsngvxj9ahsc1vm5sdrg5-protobuf-3.19.3/lib/libprotobuf.a(descriptor.o):(.text+0x1ee0f): more undefined references to `std::__throw_bad_array_new_length()' follow
       > collect2: error: ld returned 1 exit status
       > Target //build:build_wheel failed to build
       > Use --verbose_failures to see the command lines of failed build steps.
       > ERROR: /build/output/external/org_tensorflow/tensorflow/core/framework/BUILD:1400:31 Middleman _middlemen/@org_Utensorflow_S_Stensorflow_Score_Sframework_Cattr_Uvalue_Uproto_Utext-BazelCppSemantics_build_arch_k8-opt failed: (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/k8-opt/bin/external/org_tensorflow/tensorflow/tools/proto_text/gen_proto_text_functions-2.params
       > INFO: Elapsed time: 4705.455s, Critical Path: 117.50s
       > INFO: 5369 processes: 2233 internal, 3136 local.
       > FAILED: Build did NOT complete successfully
       > FAILED: Build did NOT complete successfully
       > 
       For full logs, run 'nix log /nix/store/3dx5b3blc4girz6zf772vqrsgw8457z3-bazel-build-jaxlib-0.3.0.drv'.
error: 1 dependencies of derivation '/nix/store/xb7wigdijhhfw3x3qxgl3d13ii2lzxbj-python3.9-jaxlib-0.3.0.drv' failed to build
```

[full build log](https://github.com/samuela/nixpkgs-upkeep/actions/runs/2181534947)

python3Packages.jaxlibWithCuda maintainers cc: @ndl
python3Packages.jaxlib maintainers cc: @ndl
Other cc: n/a

### Steps to reproduce
1. Checkout nixpkgs at commit 891d0226de6e281422a8d0d99814925fd7aed8d0
2. Run `nix-build -A python3Packages.jaxlib`

## Technical details
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.13.0-1021-azure, Ubuntu, 20.04.4 LTS (Focal Fossa), nobuild`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.7.0`
 - nixpkgs: `/nix/store/r902mrj2iwsjhy8gs6y0rbr9nada0z4z-source`
```

## Misc.
This issue was automatically generated by [nixpkgs-upkeep](https://github.com/samuela/nixpkgs-upkeep).
- [CI workflow](https://github.com/samuela/nixpkgs-upkeep/actions/runs/2181534947) that created this issue.
- Internal tag: 5367557120162255455

--- a/pkgs/development/compilers/cudatoolkit/versions.toml
+++ b/pkgs/development/compilers/cudatoolkit/versions.toml
@@ -44,16 +44,16 @@ gcc = "gcc9"
 version = "11.4.2"
 url = "https://developer.download.nvidia.com/compute/cuda/11.4.2/local_installers/cuda_11.4.2_470.57.02_linux.run"
 sha256 = "sha256-u9h8oOkT+DdFSnljZ0c1E83e9VUILk2G7Zo4ZZzIHwo="
-gcc = "gcc10" # can bump to 11 along with stdenv.cc
+gcc = "gcc11"
 
 ["11.5"]
 version = "11.5.0"
 url = "https://developer.download.nvidia.com/compute/cuda/11.5.0/local_installers/cuda_11.5.0_495.29.05_linux.run"
 sha256 = "sha256-rgoWk9lJfPPYHmlIlD43lGNpANtxyY1Y7v2sr38aHkw="
-gcc = "gcc10" # can bump to 11 along with stdenv.cc
+gcc = "gcc11"
 
 ["11.6"]
 version = "11.6.1"
 url = "https://developer.download.nvidia.com/compute/cuda/11.6.1/local_installers/cuda_11.6.1_510.47.03_linux.run"
 sha256 = "sha256-qyGa/OALdCABEyaYZvv/derQN7z8I1UagzjCaEyYTX4="
-gcc = "gcc10" # can bump to 11 along with stdenv.cc
+gcc = "gcc11"

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -3575,7 +3575,12 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) Security;
   };
 
-  gpu-burn = callPackage ../applications/misc/gpu-burn { };
+  gpu-burn = callPackage ../applications/misc/gpu-burn {
+    # gpu-burn doesn't build on gcc11. CUDA 11.3 is the last version to use
+    # pre-gcc11, in particular gcc9.
+    cudatoolkit = cudaPackages_11_3.cudatoolkit;
+    stdenv = gcc9Stdenv;
+  };
 
   greg = callPackage ../applications/audio/greg {
     pythonPackages = python3Packages;

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -23098,6 +23098,10 @@ with pkgs;
 
   librealsenseWithCuda = callPackage ../development/libraries/librealsense {
     cudaSupport = true;
+    # librealsenseWithCuda doesn't build on gcc11. CUDA 11.3 is the last version
+    # to use pre-gcc11, in particular gcc9.
+    cudaPackages = cudaPackages_11_3;
+    stdenv = gcc9Stdenv;
   };
 
   librealsenseWithoutCuda = callPackage ../development/libraries/librealsense {

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -4243,20 +4243,23 @@ in {
 
   jaxlib-bin = callPackage ../development/python-modules/jaxlib/bin.nix {
     cudaSupport = pkgs.config.cudaSupport or false;
-    inherit (self.tensorflow) cudaPackages;
+    # At the time of writing (2022-04-18), `cudaPackages.nccl` is broken, so we
+    # pin to `cudaPackages_11_6` instead.
+    cudaPackages = pkgs.cudaPackages_11_6;
   };
 
   jaxlib-build = callPackage ../development/python-modules/jaxlib {
     # Some platforms don't have `cudaSupport` defined, hence the need for 'or false'.
     cudaSupport = pkgs.config.cudaSupport or false;
-    inherit (self.tensorflow) cudaPackages;
+    # At the time of writing (2022-04-18), `cudaPackages.nccl` is broken, so we
+    # pin to `cudaPackages_11_6` instead.
+    cudaPackages = pkgs.cudaPackages_11_6;
   };
 
   jaxlib = self.jaxlib-build;
 
   jaxlibWithCuda = self.jaxlib-build.override {
     cudaSupport = true;
-
   };
 
   jaxlibWithoutCuda = self.jaxlib-build.override {

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -1958,7 +1958,11 @@ in {
 
   cufflinks = callPackage ../development/python-modules/cufflinks { };
 
-  cupy = callPackage ../development/python-modules/cupy { };
+  cupy = callPackage ../development/python-modules/cupy {
+    # cupy doesn't build on gcc11. CUDA 11.3 is the last version to use
+    # pre-gcc11, in particular gcc9.
+    cudaPackages = pkgs.cudaPackages_11_3;
+  };
 
   curio = callPackage ../development/python-modules/curio { };
 

